runtime/tutorials/chat_app.md

```diff
@@ -1,235 +1,303 @@
+
+export default class ChatServer {
+  private connectedClients = new Map<string, WebSocketWithUsername>();
+
+  public async handleConnection(ctx: Context) {
+    const socket = await ctx.upgrade() as WebSocketWithUsername;
+    const username = ctx.request.url.searchParams.get("username");
+
+    if (this.connectedClients.has(username)) {
+      socket.close(1008, `Username ${username} is already taken`);
+      return;
+    }
+
+    socket.username = username;
+    socket.onopen = this.broadcastUsernames.bind(this);
+    socket.onclose = () => {
+      this.clientDisconnected(socket.username);
+    };
+    socket.onmessage = (m) => {
+      this.send(socket.username, m);
+    };
+    this.connectedClients.set(username, socket);
+
+    console.log(`New client connected: ${username}`);
+  }
+
+  private send(username: string, message: any) {
+    const data = JSON.parse(message.data);
+    if (data.event !== "send-message") {
+      return;
+    }
+
+    this.broadcast({
+      event: "send-message",
+      username: username,
+      message: data.message,
+    });
+  }
+
+  private clientDisconnected(username: string) {
+    this.connectedClients.delete(username);
+    this.broadcastUsernames();
+
+    console.log(`Client ${username} disconnected`);
+  }
+
+  private broadcastUsernames() {
+    const usernames = [...this.connectedClients.keys()];
+    this.broadcast({ event: "update-users", usernames });
+
+    console.log("Sent username list:", JSON.stringify(usernames));
+  }
+
+  private broadcast(message: AppEvent) {
+    const messageString = JSON.stringify(message);
+    for (const client of this.connectedClients.values()) {
+      client.send(messageString);
+    }
+  }
+}
+```
+
+This code sets up a `handleConnection` method that is called when a new
+WebSocket connection is established. It receives a Context object from the Oak
+framework and upgrades it to a WebSocket connection. It extracts the username
+from the URL query parameters. If the username is already taken (i.e., exists in
+connectedClients), it closes the socket with an appropriate message. Otherwise,
+it sets the username property on the socket, assigns event handlers, and adds
+the socket to `connectedClients`.
+
+When the socket opens, it triggers the `broadcastUsernames` method, which sends
+the list of connected usernames to all clients. When the socket closes, it calls
+the `clientDisconnected` method to remove the client from the list of connected
+clients.
+
+When a message of type `send-message` is received, it broadcasts the message to
+all connected clients, including the sender’s username.
+
+## Build the frontend
+
+We'll build a simple UI that shows a text input and a send button and displays
+the sent messages, alongside a list of users in the chat.
+
+### HTML
+
+In your new project directory, create a `public` folder and add an `index.html`
+file and add the following code:
+
+```html title="index.html"
```

Comment:
Let's add the doctype to be fully valid html code.

````suggestion
```html title="index.html"
<!DOCTYPE html>
````